### ISSUE TITLE
Request for AF_PACKET along with raw socket(SOCK_RAW) support.

### ISSUE THREAD

**Issue by skris14** [gramineproject/gramine#633]
## Description of the problem
Currently Gramine does not support AF_PACKET socket type.
For one of the application that we are running in Gramine, it
needs AF_PACKET socket along with SOCK_RAW (socket_type).
Currently, it looks like Gramine does not seem to support 
raw socket(SOCK_RAW).

## Steps to reproduce
Any C application code, that invokes below:

int fd = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
if (fd == -1) {
    perror("socket");
    return -1;
}

## Actual results
Does not open the socket type, since it is NOT supported in Gramine.

## Additional information
AF_PACKET socket along with raw socket support(SOCK_RAW)  is needed for running UPF/BESS  application(https://github.com/omec-project/upf ) in Gramine.

## Gramine commit hash
commit 1b83451a1e8e38e7d56c8e6b2fe923504a4dc6a0

**Comment by skris14** [comment]
## Steps to reproduce
Any C application code, that invokes below:
'''int fd = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ALL));
if (fd == -1) {
    perror("socket");
    return -1;
}'''

**Comment by boryspoplawski** [comment]
@skris14 Could you elaborate more on what the workload does? From a quick look it seems to be a generic framework for managing network packets. It that's the case why would you want to run that inside an enclave?

**Comment by skris14** [comment]
@boryspoplawski , UPF(user-plane function, (https://github.com/omec-project/upf )) can get deployed in public cloud enviornment. So, by running the workload in SGX, we are protecting integrity of the code(by running it within SGX), in an untrusted public cloud deployment.

**Comment by skris14** [comment]
@boryspoplawski , like I have tried to explain, we are NOT trying to protect the pay-load, that is the outside scope. We are ONLY trying to run the code within SGX for the following reason-> UPF(user-plane function, (https://github.com/omec-project/upf )) can get deployed in public cloud enviornment. So, by running the workload in SGX, we are protecting integrity of the code(by running it within SGX), in an untrusted public cloud deployment.

**Comment by boryspoplawski** [comment]
Let me rephrase my question: what is stopping the "public cloud environment" from modifying the output from the enclave (network packets)?

**Comment by skris14** [comment]
@boryspoplawski , that is outside the scope, and we rely on end-to-end security like TLS between the 2 end-points. This UPF module is just like a router, that sits in-between two end-points.

**Comment by boryspoplawski** [comment]
>  we rely on end-to-end security like TLS between the 2 end-points. This UPF module is just like a router, that sits in-between two end-points.

Nice! Then there is no reason in putting that "router" inside an enclave. If the malicious OS can modify the runtime behavior of that router, then it also can modify the packets as they leave. Putting the code inside the enclave protects from nothing in this scenario.

**Comment by skris14** [comment]
Hi Borys, I have tried to explain, that we are NOT protecting the payload, couple of times already. We are only protecting integrity of code by running in UPF in SGX, and with the requirement that end-to-end security be provided thro mechansims like TLS. We also protect security sensitive data managed by UPF, by using hash-maps in UPF, and these hash-maps get allocated in SGX memory. Hope it is clear.

**Comment by skris14** [comment]
@boryspoplawski , are you available in slack, I can discuss via slack, if that helps to explain. Also Mona is familiar with the security requirements for this UPF work, so you can also check with her, if you have any further questions.

**Comment by skris14** [comment]
Just to clarify, this support (AF_PACKET) is missing  in Gramine. It shouldnt matter which application needs it.

**Comment by boryspoplawski** [comment]
> I have tried to explain, that we are NOT protecting the payload, couple of times already. 

Yes, I know, I got it the first time.

> We are only protecting integrity of code by running in UPF in SGX, and with the requirement that end-to-end security be provided thro mechansims like TLS.

And there is no point in protecting the code if you have no way of verifying the output was not modified.
If this "router" sits in-between two end-points, as you stated, then it's no difference if it's code runs inside an enclave - nothing it does needs protection, since all of it can be changed by malicious host OS anyway.

> We also protect security sensitive data managed by UPF, by using hash-maps in UPF, 

hash-maps as in https://en.wikipedia.org/wiki/Hash_table? This makes no sense.

> are you available in slack, I can discuss via slack,

What slack?

**Comment by boryspoplawski** [comment]
> Just to clarify, this support (AF_PACKET) is missing in Gramine. It shouldnt matter which application needs it.

It very much does matter, we are not going to spend time implementing a feature (a very complicated one as well) if it doesn't make sense to have it.

**Comment by mkow** [comment]
@skris14: I'm sorry, but your description is really confusing and also doesn't make much sense to me.

> we rely on end-to-end security like TLS between the 2 end-points
> end-to-end security be provided thro mechansims like TLS
> This UPF module is just like a router, that sits in-between two end-points.

Ok, so the code inside the enclave doesn't look inside the packets, it's only routing them? Then as Borys said, why is it running inside the enclave? You gain exactly nothing by using an enclave for this, because the routing may be altered by the host anyways, and the code running inside the enclave won't even notice.

> Just to clarify, this support (AF_PACKET) is missing in Gramine. It shouldnt matter which application needs it.

It _does_ matter. There are hundreds of Linux subsystems not supported in Gramine, e.g. video drivers, but this doesn't mean that we plan to implement them someday. We only implement subsystems for which there exists a valid use-case inside TEEs.

**Comment by skris14** [comment]
@boryspoplawski Sorry, if it is still not clear, the only way forward is to have a meeting and discuss. I can try to have a meeting with you.

**Comment by skris14** [comment]
@mkow This is a valid use-case, in our opinion. You can also talk to Mona, since she is aware of the security requirements for this use-case.

**Comment by mkow** [comment]
@skris14: Instead of doing a meeting (which would consume a lot of our time) please just clearly describe what are you doing and your security model. I'm closing this until we get a clearer design from you (but this really seems like a by-design insecure use of SGX).

**Comment by monavij** [comment]
@mkow  @boryspoplawski   The use case involves protecting UPF (User plane Function) VNF that receives user subscriber data and routes packets based on certain subscriber policies. So while we are not trying to protect the packet data we are trying to protect the private subscriber information and want to make sure no one can tamper with that process, so protecting the integrity of execution.

**Comment by mkow** [comment]
> no one can tamper with that process

But the host will always be able to tamper with it, the destinations of packets leaving the enclave can't be protected - they are both in clear text and modifiable, because the host and network devices need to know what to do with them.

**Comment by monavij** [comment]
Yes so it depends on the threat model. DOS attacks can't be prevented as host is in control and they cna choose to not send packets etc and there are other ways to mitgate or detect such attacks. But the use case is to protect the privacy and confidentiality of sensitive subscriber info which can be leaked if UPF is not running in a TEE

**Comment by boryspoplawski** [comment]
> DOS attacks can't be prevented as host is in control and they cna choose to not send packets etc and there are other ways to mitgate or detect such attacks. 

We are not talking about DOS attacks.

> But the use case is to protect the privacy and confidentiality of sensitive subscriber info which can be leaked if UPF is not running in a TEE

And what that sensitive info would be? Like routing tables? Ignoring the question why such info is sensitive, they could be inferred with the workload running in a TEE (e.g. log and match input to output packets and then you can deduce routing).

**Comment by skris14** [comment]
Had an internal conversation with colleague (Michael), and one thing, that I missed to mention in this discussion, is that as per 3GPP specification, the traffic flowing thro UPF will have protection through IPSec ESP(confidentiality, integrity and more), which should help mitigate the security concerns.

**Comment by boryspoplawski** [comment]
> the traffic flowing thro UPF will have protection through IPSec ESP

Like a tunnel between two endpoints and UPF just acting as a middleman or multiple tunnels terminated in UPF?

**Comment by woju** [comment]
> I would recommend offline email/chat/meeting

You're welcome to write at our mailinglist (users-at-gramineproject.io), or we have a gitter channel, https://gitter.im/gramineproject/community, which you can also access from matrix as https://matrix.to/#/#gramineproject_community:gitter.im, but the final design of this RFC/feature request and (if accepted) subsequent review of the implementation will necessarily happen here on github.

**Comment by skris14** [comment]
there is ipsec tunnel between the radio-tower and UPF.
